File: packages/sdk-ts/src/sanitizer.ts
export interface InjectionSanitizeResult {
  detected: boolean;
  type?: 'INDIRECT_PROMPT_INJECTION' | 'ZERO_WIDTH_UNICODE' | 'OBFUSCATED_PAYLOAD' | 'DESTRUCTIVE_PATTERN';
  patternMatched?: string;
}

export class InjectionSanitizer {
  private static INDIRECT_INJECTION_PATTERNS = [
    /\[SYSTEM\s*OVERRIDE\]/i,
    /IGNORE\s+(ALL\s+)?PREVIOUS\s+(INSTRUCTIONS|GOALS|RULES)/i,
    /DISREGARD\s+PRIOR\s+(RULES|GOALS|INSTRUCTIONS)/i,
    /YOU\s+ARE\s+NOW\s+IN\s+DAN\s+MODE/i,
    /UNRESTRICTED\s+(MODE|PROMPT)/i,
    /NEW\s+SYSTEM\s+PROMPT:/i,
    /ADMIN_OVERRIDE_KEY/i,
  ];

  private static DESTRUCTIVE_PATTERNS = [
    /DROP(?:\/\*[\s\S]*?\*\/|\s)+TABLE/i,
    /DELETE(?:\/\*[\s\S]*?\*\/|\s)+FROM(?:\/\*[\s\S]*?\*\/|\s)+[a-z_]+/i,
    /TRUNCATE(?:\/\*[\s\S]*?\*\/|\s)+TABLE/i,
    /UNION(?:\/\*[\s\S]*?\*\/|\s)+SELECT/i,
    /INFORMATION_SCHEMA/i,
    /rm\s+-rf\s+/i,
    /chmod\s+777/i,
    /mkfs\./i,
  ];

  static inspect(payloadStr: string): InjectionSanitizeResult {
    let unescapedPayload = payloadStr;
    
    // 0. URL-Decode payload if encoded (%5BSYSTEM%20OVERRIDE%5D)
    try {
      if (unescapedPayload.includes('%')) {
        unescapedPayload = decodeURIComponent(unescapedPayload);
      }
    } catch {
      // Ignore if malformed URI component
    }

    // 0b. Unicode NFKC Normalization (converts full-width ＤＡＮ ＭＯＤＥ to DAN MODE)
    const normalizedPayload = unescapedPayload
      .normalize('NFKC')
      .replace(/\\t|\\n|\\r/g, ' ');

    // 1. Check Zero-Width Unicode Characters (used to hide injections)
    const zeroWidthRegex = /[\u200B-\u200D\uFEFF]/;
    if (zeroWidthRegex.test(normalizedPayload)) {
      return {
        detected: true,
        type: 'ZERO_WIDTH_UNICODE',
        patternMatched: 'Hidden Zero-Width Unicode Characters Detected',
      };
    }

    // 2. Check Indirect Prompt Injection Patterns
    for (const pattern of this.INDIRECT_INJECTION_PATTERNS) {
      if (pattern.test(normalizedPayload)) {
        return {
          detected: true,
          type: 'INDIRECT_PROMPT_INJECTION',
          patternMatched: pattern.source,
        };
      }
    }

    // 3. Check Destructive Patterns
    for (const pattern of this.DESTRUCTIVE_PATTERNS) {
      if (pattern.test(normalizedPayload)) {
        return {
          detected: true,
          type: 'DESTRUCTIVE_PATTERN',
          patternMatched: pattern.source,
        };
      }
    }

    // 4. Base64 Obfuscation Inspection
    const base64Regex = /([A-Za-z0-9+/]{8,}={0,2})/g;
    let match;
    while ((match = base64Regex.exec(normalizedPayload)) !== null) {
      try {
        const decoded = Buffer.from(match[1], 'base64').toString('utf-8').normalize('NFKC');
        for (const pattern of [...this.INDIRECT_INJECTION_PATTERNS, ...this.DESTRUCTIVE_PATTERNS]) {
          if (pattern.test(decoded)) {
            return {
              detected: true,
              type: 'OBFUSCATED_PAYLOAD',
              patternMatched: `Base64 Decoded: ${pattern.source}`,
            };
          }
        }
      } catch {
        // Ignored if not valid Base64
      }
    }

    return { detected: false };
  }
}

File: packages/sdk-rust/src/sanitizer.rs
use regex::Regex;
use base64::{Engine as _, engine::general_purpose};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InjectionSanitizeResult {
    pub detected: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pattern_matched: Option<String>,
}

impl Default for InjectionSanitizeResult {
    fn default() -> Self {
        Self {
            detected: false,
            r#type: None,
            pattern_matched: None,
        }
    }
}

pub struct InjectionSanitizer;

impl InjectionSanitizer {
    fn indirect_injection_patterns() -> Vec<Regex> {
        vec![
            Regex::new(r"(?i)\[SYSTEM\s*OVERRIDE\]").unwrap(),
            Regex::new(r"(?i)IGNORE\s+ALL\s+PREVIOUS\s+INSTRUCTIONS").unwrap(),
            Regex::new(r"(?i)DISREGARD\s+PRIOR\s+RULES").unwrap(),
            Regex::new(r"(?i)YOU\s+ARE\s+NOW\s+IN\s+DAN\s+MODE").unwrap(),
            Regex::new(r"(?i)NEW\s+SYSTEM\s+PROMPT:").unwrap(),
            Regex::new(r"(?i)ADMIN_OVERRIDE_KEY").unwrap(),
        ]
    }

    fn destructive_patterns() -> Vec<Regex> {
        vec![
            Regex::new(r"(?i)DROP\s+TABLE").unwrap(),
            Regex::new(r"(?i)DELETE\s+FROM\s+[a-z_]+").unwrap(),
            Regex::new(r"(?i)TRUNCATE\s+TABLE").unwrap(),
            Regex::new(r"(?i)rm\s+-rf\s+").unwrap(),
            Regex::new(r"(?i)chmod\s+777").unwrap(),
            Regex::new(r"(?i)mkfs\.").unwrap(),
        ]
    }

    fn zero_width_regex() -> Regex {
        Regex::new(r"[\u200B-\u200D\uFEFF]").unwrap()
    }

    fn base64_regex() -> Regex {
        Regex::new(r"([A-Za-z0-9+/]{8,}={0,2})").unwrap()
    }

    pub fn inspect(payload: &str) -> InjectionSanitizeResult {
        let normalized = payload.replace("\\t", " ").replace("\\n", " ").replace("\\r", " ");

        if Self::zero_width_regex().is_match(&normalized) {
            return InjectionSanitizeResult {
                detected: true,
                r#type: Some("ZERO_WIDTH_UNICODE".to_string()),
                pattern_matched: Some("Hidden Zero-Width Unicode Characters Detected".to_string()),
            };
        }

        for pattern in Self::indirect_injection_patterns() {
            if pattern.is_match(&normalized) {
                return InjectionSanitizeResult {
                    detected: true,
                    r#type: Some("INDIRECT_PROMPT_INJECTION".to_string()),
                    pattern_matched: Some(pattern.as_str().to_string()),
                };
            }
        }

        for pattern in Self::destructive_patterns() {
            if pattern.is_match(&normalized) {
                return InjectionSanitizeResult {
                    detected: true,
                    r#type: Some("DESTRUCTIVE_PATTERN".to_string()),
                    pattern_matched: Some(pattern.as_str().to_string()),
                };
            }
        }

        for captures in Self::base64_regex().captures_iter(&normalized) {
            if let Some(m) = captures.get(1) {
                if let Ok(decoded_bytes) = general_purpose::STANDARD.decode(m.as_str()) {
                    if let Ok(decoded) = String::from_utf8(decoded_bytes) {
                        for pattern in Self::indirect_injection_patterns().into_iter().chain(Self::destructive_patterns()) {
                            if pattern.is_match(&decoded) {
                                return InjectionSanitizeResult {
                                    detected: true,
                                    r#type: Some("OBFUSCATED_PAYLOAD".to_string()),
                                    pattern_matched: Some(format!("Base64 Decoded: {}", pattern.as_str())),
                                };
                            }
                        }
                    }
                }
            }
        }

        InjectionSanitizeResult::default()
    }
}